src/transaction/log_replication_atomic.cpp

```diff
@@ -402,37 +402,72 @@ namespace cublog
 	assert_release (false);
       }
 
-#if !defined (NDEBUG)
-    /* there will be no duplicated object in the map */
-    auto [begin, end] = m_locked_classes.equal_range (trid);
-    if (std::count_if (begin, end, [classoid] (const auto &it)
-    {
-      return it.second == *classoid;
-    }) > 0)
-    {
-      assert (false);
-    }
-#endif
+    m_locked_classes[trid].emplace (*classoid);
+  }
+
+  void
+  atomic_replicator::bookkeep_classname (cubthread::entry &thread_entry, const OID *classoid)
+  {
+    char *classname = NULL;
+
+    assert (!OID_ISNULL (classoid) && !OID_ISTEMP (classoid));
 
-    m_locked_classes.emplace (trid, *classoid);
+    (void) heap_get_class_name (&thread_entry, classoid, &classname);
+    if (classname != NULL)
+      {
+	m_classname_map.emplace (*classoid, classname);
+	free_and_init (classname);
+      }
+    else
+      {
+	/* if classname == NULL, then it is CREATE TABLE/VIEW case */
+      }
+  }
+
+  void
+  atomic_replicator::update_classname_cache_for_ddl (cubthread::entry &thread_entry, const OID *classoid)
+  {
+    /* update locator_Mht_classnames only if needed */
+    char *classname = NULL;
```

Comment:
Just to make it clear for me - to make sure I understand:
- the class name obtained from `heap_get_class_name` here is the actual class name obtained from the heap page and is the actual result __after__ replicating the DDL-related log records
- the class name in `m_classname_map` is the class as obtained from `heap_get_class_name` __before__ replicating the DDL-related log records - in `bookkeep_classname` above.

And based on this, below it is decided if it is a CREATE/DROP/UPDATE statement.

Question:
Are we sure that - for `CREATE` statements - the class name does not appear somewhere in the log records chain as as a result of multiple `LOG_SCHEMA_MODIFICATION_LOCK` for a `CREATE ...` statement? Or is this possibility covered by-design by the `locator_update_classname_entry` function.

In other words:
Is it possible for a `CREATE` statement to not be detected as because the __before__ (in  `m_classname_map`) and __after__ (directly from the heap) classnames contain the same name?

Reply:
Thanks..! I found a hole.. I will find out some general solution for this
possible scenario : 

CREATE TABLE tbl ( a int);
ALTER TABLE tbl add column b int;
ALTER TABLE tbl add column b int; -- [semantic check error], insert "tbl" into `m_classname_map` 
COMMIT; -- before and after can contain the same name..

Reply:
One idea would be to only take into account the first occurence of the `LOG_SCHEMA_MODIFICATION_LOCK` (as, in your example - resulting from the first `ALTER TABLE` statement).
Upon a second time, the fact that the `OID` points to a valid class name is not taken into account.

As in, for a transaction occuring in the log record chain:
1. `LOG_SCHEMA_MODIFICATION_LOCK` encountered, class name not found - record this information and the fact that we're dealing with a `CREATE`
2. `LOG_SCHEMA_MODIFICATION_LOCK` encountered again, do not check anymore for the class name
3. at `COMMIT` - the logic should work and properly detect a `CREATE` statement

This is more of a "Captain Obvious" remark, as I am sure you have already a solution for this. :)